### DISCUSSION TITLE
Is there anyway to select key exchange algorithm in TLS 1.3?

### DISCUSSION THREAD

**Discussion by afshinpir** [openssl/openssl#24922] [Q&A] [answered]
Hi all,

I know that key exchange algorithm is agreed upon in TLSv1.3. But I wonder if there is a way to limits allowed algorithms for key exchange and authentication or not. In addition, is there anyway to see what key exchange algorithm we negotiated upon? Because if I try something like `SSL_CIPHER_get_kx_nid(SSL_get_current_cipher(ssl))`, I will get kx_any as result.

Any idea?

**Comment by t8m** [comment, marked as answer]
With the exception of the psk-kex, which is off by default and is controlled by SSL_options SSL_OP_ALLOW_NO_DHE_KEX and SSL_OP_PREFER_NO_DHE_KEX, the concrete key exchange type is controlled by the Group setting (SSL_CTX_set1_groups_list()). I.e., you do not list FFDHE groups there if you do not want to allow _classic_ DH.

**Comment by afshinpir** [comment]
@t8m 
So if I use group "P-384", the only possible key exchange algorithm will be ECDHE and authentication is ECDSA? It is sad that we cannot get the concrete negotiated algorithm when negotiation is done. `SSL_CIPHER_description()` will always return _'any'_ for key exchange and authentication algorithms.

**Reply by t8m** [reply]
Authentication can be not only ECDSA but also RSA - depending on the server's certificate. Allowed authentication is controlled by the SignatureAlgorithms setting. But yes, the key exchange will be ECDHE.

Ciphers in TLSv1.3 are completely dissociated from the key exchange and authentication so SSL_CIPHER_description() cannot really return anything else than any. 

Which authentication and which key exchange was selected can be queried by SSL_get(_peer)_signature(_type)_nid() and SSL_get_negotiated_group().
